### PR TITLE
dockerTools: add docs to docker build functions

### PR DESCRIPTION
**Pull request by adnelson**
###### Motivation for this change

The current docker functionality is useful but dense in content while sparse in documentation. This adds documentation and a few stylistic improvements. In addition, it fixes a shortcoming which is the hard-coding of the `/nix/store` store directory, replacing it with a parameter.

There's more that could be done on this front, especially higher-level documentation, but it's a little better than it was before I feel. 
###### Things done
- Added documentation
- A few stylistic tweaks (indentation etc)
- Ability to set the nix store directory to support non-standard nix store prefixes
- [ ] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
  - [X] NixOS
  - [ ] OS X
  - [X] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by tboettch** [comment]
I'm not able to build your branch locally (same issue as Travis) but reviewing the diff it looks like it rolls back the work I did on #18896 to remove redundancy between image layers? I posted a simple test case on that PR; could you check to see if glibc is still being properly omitted from the second layer inside the image tarball?

I'm not especially attached to the implementation I used in that PR (I'm a bit of Bash amateur) but the de-duplication feature is reasonably important to me.

**Comment by adnelson** [comment]
Oh, I'm sorry, yeah this was made off of an earlier branch that I had. I'll fix that.

**Comment by adnelson** [comment]
@tboettch Sorry about that, I added it back in. I changed the indentation on your function to two spaces, just in line with the other stylistic things in this PR :)

**Comment by Mic92** [comment]
Can you squash the first 3 commits into one?

File: pkgs/build-support/docker/default.nix [hunk 1]
```diff
@@ -23,109 +41,134 @@ rec {
...
-account sufficient pam_unix.so
-auth sufficient pam_rootok.so
-password requisite pam_unix.so nullok sha512
-session required pam_unix.so
-EOF
+    account sufficient pam_unix.so
+    auth sufficient pam_rootok.so
+    password requisite pam_unix.so nullok sha512
+    session required pam_unix.so
+    EOF
```
Comment (cstrahan) [inline pkgs/build-support/docker/default.nix]: Won't these leading spaces be included in the file? And we don't want that, right?

Comment (adnelson) [inline pkgs/build-support/docker/default.nix]: No, the nix parser looks for the left-most indentation in a `''` string and uses that as the zeroth column in the string. Putting the heredoc at indentation 0 in the nix expression has the effect of indenting everything else.

Comment (adnelson) [inline pkgs/build-support/docker/default.nix]: For example:

```nix
with import <nixpkgs> {};

      writeScript "test" ''
        echo "hello"
        cat <<EOF
this is some text
here it is
EOF
        echo "hi"
      ''
```

After building this:

```bash
$ cat result 
        echo "hello"
        cat <<EOF
this is some text
here it is
EOF
        echo "hi"
```

On the other hand, if your expression is:

```nix
with import <nixpkgs> {};

      writeScript "test" ''
        echo "hello"
        cat <<EOF
        this is some text
        here it is
        EOF
        echo "hi"
      ''
```

Then after building it:

```bash
$ cat result
echo "hello"
cat <<EOF
this is some text
here it is
EOF
echo "hi"
```

**Comment by adnelson** [comment]
Squashed

File: pkgs/build-support/docker/default.nix (hunk 1 above)
Comment (cstrahan) [inline pkgs/build-support/docker/default.nix]: Oh, good call. I knew about the omission of common left-most indentation, but I wasn't paying close enough attention to the surrounding context / indentation. :+1:

**Review by cstrahan — APPROVED** [review]
Looks good :+1:

**Comment by adnelson** [comment]
Not sure what's going on with this error:

```
./maintainers/scripts/travis-nox-review-pr.sh: line 20: nix-build: command not found
```

**Comment by Mic92** [comment]
there was a hick-up on travis. No need to worry. We test this stuff locally anyways.

**Review by domenkozar — APPROVED** [review]

**Comment by Mic92** [comment]
do you have an minimal example, how I would create such and run container in docker?

**Comment by adnelson** [comment]
@Mic92 you can use `buildImage` to make a derivation which builds a tarball. After you build the derivation you can pass the result to `docker load`.

**Comment by Mic92** [comment]
thanks. I used `docker import` instead of `docker load`:

I stumbled over this gist: https://gist.github.com/rvl/c1ca657882a883195f49
If we include these examples, it would be better then nothing.

**Comment by adnelson** [comment]
Good find, those look like great examples. I'll assume they're free to use by the author and add them to my PR.

**Comment by adnelson** [comment]
@Mic92 I added that file (with a few tweaks) as an `examples.nix` file alongside the docker tools. However, some of the examples don't build -- not sure what's going on with that. If it's not a result of my changes, I think we can leave it as-is, but I'll take a look and make sure (or possibly fix it if it is broken).

**Comment by adnelson** [comment]
Hmm, so unfortunately it appears that expressions which define a `runAsRoot` cannot be run if virtualbox is running. This is a bummer -- global state FTL.

**Comment by adnelson** [comment]
Maybe there's some way around it though.

**Comment by adnelson** [comment]
So I was getting some permission denied errors because folders have the potential to overwrite each other during the copy contents phase; I switched to using `rsync` instead which seems like a good upgrade over `cp` anyway. As to the `pullImage` function, it seems to be broken; either the API has changed or we're using it wrong. Oh well :(

**Comment by cstrahan** [comment]
> As to the pullImage function, it seems to be broken; either the API has changed or we're using it wrong.

We might have some updating to do:

* https://github.com/docker/docker/blob/master/image/spec/v1.2.md
* https://docs.docker.com/registry/spec/api/ (at present, the V2 registry spec)
* https://github.com/docker/docker/pull/11271 (Push and pull to v2 registries with sha256 instead of tarsum)

**Comment by adnelson** [comment]
Also I just found that there's what looks to be a pretty full-featured python library for docker; probably if we used this library instead of curls we'd be able to get the functionality in a cleaner way

**Comment by adnelson** [comment]
@cstrahan are you suggesting modifying how we build the layers themselves? That might be a great idea but seeing as the current stuff seems to work OK (not that I've used it a whole ton yet), it might not be necessary at this point.

**Comment by cstrahan** [comment]
> Also I just found that there's what looks to be a pretty full-featured python library for docker; probably if we used this library instead of curls we'd be able to get the functionality in a cleaner way

If you're referring to https://github.com/docker/docker-py, I think that only provides a client for communicating with the docker daemon, and not connecting to the registry directly: https://docker-py.readthedocs.io/en/latest/api/

> @cstrahan are you suggesting modifying how we build the layers themselves? That might be a great idea but seeing as the current stuff seems to work OK (not that I've used it a whole ton yet), it might not be necessary at this point.

Not for now, but maybe later.

**Comment by adnelson** [comment]
Ah, well that's too bad. I guess we'll have to just figure out the proper API calls and update the curls.

I guess I'm done with this then; if no one else has comments we could merge it

**Comment by Mic92** [comment]
@adnelson I guess we could merge this for now?

**Comment by adnelson** [comment]
I don't have merge power; I'm waiting for someone else to hit the button...

**Comment by Mic92** [comment]
Thanks. Good work!

**Pull request merged**